zip-0312.rst

```diff
@@ -1,7 +1,466 @@
+    represent a Scalar in the range \[0, `G.Order()` - 1\].
+
+- Hash (`H`): BLAKE2b-512 [#BLAKE]_ (BLAKE2b with 512-bit output and 16-byte personalization string),
+  and Nh = 64.
+
+  - H1(m): Implemented by computing BLAKE2b-512("FROST_RedPallasR", m), interpreting
+    the 64 bytes as a little-endian integer, and reducing the resulting integer
+    modulo L = 0x40000000000000000000000000000000224698fc0994a8dd8c46eb2100000001.
+  - H2(m): Implemented by computing BLAKE2b-512("Zcash_RedPallasH", m), interpreting
+    the 64 bytes as a little-endian integer, and reducing the resulting integer
+    modulo L = 0x40000000000000000000000000000000224698fc0994a8dd8c46eb2100000001.
+    (This is equivalent to :math:`\mathsf{H}^\circledast(m)`, as defined in
+    [#protocol-concretereddsa]_ parametrized with [#protocol-pallasandvesta]_.)
+  - H3(m): Implemented by computing BLAKE2b-512("FROST_RedPallasN", m), interpreting
+    the 64 bytes as a little-endian integer, and reducing the resulting integer
+    modulo L = 0x40000000000000000000000000000000224698fc0994a8dd8c46eb2100000001.
+  - H4(m): Implemented by computing BLAKE2b-512("FROST_RedPallasM", m).
+  - H5(m): Implemented by computing BLAKE2b-512("FROST_RedPallasC", m).
+
+
+Reference implementation
+========================
+
+TODO: add links to implementation
+
+
+References
+==========
+
+.. [#BLAKE] `BLAKE2: simpler, smaller, fast as MD5 <https://blake2.net/#sp>`_
+.. [#RFC2119] `RFC 2119: Key words for use in RFCs to Indicate Requirement Levels <https://www.rfc-editor.org/rfc/rfc2119.html>`_
+.. [#FROST] `Draft RFC: Two-Round Threshold Schnorr Signatures with FROST <https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-08.html>`_
+.. [#frost-protocol] `Draft RFC: Two-Round Threshold Schnorr Signatures with FROST. Section 5: Two-Round FROST Signing Protocol <https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-08.html#section-5>`_
+.. [#frost-removingcoordinator] `Draft RFC: Two-Round Threshold Schnorr Signatures with FROST. Section 7.3: Removing the Coordinator Role <https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-08.html#section-7.3>`_
+.. [#frost-primeordergroup] `Draft RFC: Two-Round Threshold Schnorr Signatures with FROST. Section 3.1: Prime-Order Group <https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-08.html#section-3.1>`_
+.. [#frost-tdkg] `Draft RFC: Two-Round Threshold Schnorr Signatures with FROST. Appendix B: Trusted Dealer Key Generation <https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-08.html#appendix-B>`_
+.. [#frost-randomscalar] `Draft RFC: Two-Round Threshold Schnorr Signatures with FROST. Appendix C: Random Scalar Generation <https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-08.html#appendix-C>`_
+.. [#protocol-concretereddsa] `Zcash Protocol Specification, Version 2022.3.4 [NU5]. Section 5.4.7: RedDSA, RedJubjub, and RedPallas <https://protocol/protocol.pdf#concretereddsa>`_
```

Comment:
TODO(ZIP Editors): update these to the latest version of the spec before publishing.